src/io/cached_itarbundle.rs

```diff
@@ -0,0 +1,693 @@
+             this most probably is not Tectonic's fault \
+             -- please check your network connection.",
+            name
+        );
+    } else if any_failed {
+        tt_note!(status, "download succeeded after retry");
+    }
+
+    Ok(buf)
+}
+
+fn parse_index_line(line: &str) -> Result<Option<(String, FileInfo)>> {
+    let mut bits = line.split_whitespace();
+
+    if let (Some(name), Some(offset), Some(length)) = (bits.next(), bits.next(), bits.next()) {
+        Ok(Some((
+            name.to_owned(),
+            FileInfo {
+                offset: offset.parse::<u64>()?,
+                length: length.parse::<u64>()?,
+            },
+        )))
+    } else {
+        // TODO: preserve the warning info or something!
+        Ok(None)
+    }
+}
+
+/// Attempts to find the redirected url, download the index and digest.
+fn get_everything(url: &str, status: &mut dyn StatusBackend) -> Result<(String, String, String)> {
+    let url = resolve_url(url, status)?;
+
+    let index = {
+        let mut index = String::new();
+        get_index(&url, status)?.read_to_string(&mut index)?;
+        index
+    };
+
+    let digest_text = {
+        // Find the location of the digest file.
+        let digest_info = {
+            let mut digest_info = None;
+            for line in index.lines() {
+                if let Some((name, info)) = parse_index_line(line)? {
+                    if name == digest::DIGEST_NAME {
+                        digest_info = Some(info);
+                        break;
+                    }
+                }
+            }
+            ctry!(digest_info; "backend does not provide needed {} file", digest::DIGEST_NAME)
+        };
+
+        let mut range_reader = HttpRangeReader::new(&url);
+        String::from_utf8(get_file(
+            &mut range_reader,
+            digest::DIGEST_NAME,
+            digest_info.offset,
+            digest_info.length as usize,
+            status,
+        )?)
+        .map_err(|e| e.utf8_error())?
+    };
+
+    Ok((digest_text, index, url))
+}
+
+/// Load cached data.
+///
+/// If any of the files is not found return None.
+fn load_cache(
+    digest_path: &Path,
+    redirect_base: &Path,
+    index_base: &Path,
+) -> Result<Option<(String, String, HashMap<String, FileInfo>)>> {
+    // Convert file-not-found errors into None.
+    match load_cache_inner(digest_path, redirect_base, index_base) {
+        Ok(r) => Ok(Some(r)),
+        Err(Error(ErrorKind::Io(ref e), _)) if e.kind() == IoErrorKind::NotFound => Ok(None),
+        Err(e) => Err(e),
+    }
+}
+
+/// See `load_cache`.
+fn load_cache_inner(
+    digest_path: &Path,
+    redirect_base: &Path,
+    index_base: &Path,
+) -> Result<(String, String, HashMap<String, FileInfo>)> {
+    let digest_text = {
+        let f = File::open(digest_path)?;
+        let mut digest_text = String::with_capacity(64);
+        f.take(64).read_to_string(&mut digest_text)?;
```

Comment:
I'd prefer if the magic number here were expressed as a `const usize` at the top of the file.